### PR TITLE
Fix Document.initAction to fire when opening from the IDE

### LINKED ISSUE #4549: Document.initAction not called from the editor
## Environment

* SuperCollider version: 3.10.0
* Operating system: Win 10 64b

## Steps to reproduce

```supercollider
Document.initAction = { "initAction called".postln; }
```

Then `Ctrl+N`.

## Expected vs. actual behavior

Message is shown when using `Document.new` and `Document.open`, but not when using the File menu or keyboard shortcuts. However the latter cases do create documents, because `Document.allDocuments` changes. Therefore I expect that `initAction` is called in all these cases.

### PR DESCRIPTION
**Pull request by jamshark70**
## Purpose and Motivation

Fixes #4549.

Previously, `Document.initAction = { ... }` or autorun documents took effect only when opening the document programmatically (`Document.open` or `Document.new`), but these were ignored when opening the document using IDE menu commands.

The simplest fix also means that autoRun documents will execute whenever recompiling the class library (if they are open). Per discussion under #4549, this seems to match the original autoRun behavior under Cocoa, and it's justifiable as an easy way to write "soft" startup code. However, the documentation inaccurately claimed that autoRun fires only when opening the document. So I updated the help file as well.

This PR touches two behaviors that are impossible to test automatically:

- User action in the IDE.
- Startup behavior after recompiling the library (as this would halt the entire test suite).

I've added tests for initAction and autoRun when opening a document programmatically. That's about all I think I can do.

## Types of changes

- Documentation
- Bug fix

## To-do list

- [x] Code is tested
- [x] All tests are passing
- [x] Updated documentation
- [x] This PR is ready for review

### CODE REVIEW

File: testsuite/classlibrary/TestDocument.sc [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+				Document.initAction = { success = true };
+				doc = Document.new;
+				0.1.wait;
+				doc.close;
+			} {
+				Document.initAction = save;
+			};
+			this.assert(success, "Document.new should fire Document.initAction function");
+		} {
+			this.assert(false, "Cannot test Document.new behavior in a non-ScIDE session");
```
Comment (snappizz) [inline testsuite/classlibrary/TestDocument.sc]: UnitTest doesn't have a built-in mechanism for marking a test as "skip", but borrowing semantics from other test frameworks, i would expect this to pass, not fail. if we put an "assert false" in here, the test suite will always fail if you are running in an environment without the IDE, which doesn't seem right to me.

i'd suggest replacing this assertion with an empty code block with a comment like `// TODO: skip this test when UnitTest has skip functionality`

File: testsuite/classlibrary/TestDocument.sc [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	classvar <>success;
+
+	test_new_document_runs_initAction {
+		var success = false, doc, save;
+		if(Platform.ideName == "scqt") {
+			save = Document.initAction;
+			protect {
+				Document.initAction = { success = true };
+				doc = Document.new;
+				0.1.wait;
```
Comment (snappizz) [inline testsuite/classlibrary/TestDocument.sc]: `0.1.wait` in a test is a code smell. can you justify why this is necessary?

File: testsuite/classlibrary/TestDocument.sc [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+			this.assert(false, "Cannot test Document.new behavior in a non-ScIDE session");
+		};
+	}
+
+	test_open_autorun_document_runs_code {
+		var path = PathName.tmp +/+ "autoRunTest%.scd".format(UniqueID.next),
+		file, doc;
+		if(Platform.ideName == "scqt") {
+			if(File.exists(path)) {
+				this.assert(false, "Document autoRun test file path should not exist: '%'".format(path));
```
Comment (snappizz) [inline testsuite/classlibrary/TestDocument.sc]: would raising an exception make more sense here than `assert false`? it's not the thing we're really testing here.

File: testsuite/classlibrary/TestDocument.sc (hunk 1 above)
Comment (jamshark70) [reply testsuite/classlibrary/TestDocument.sc]: Ok, fair enough. Should there or should there not be some record of tests that were skipped, and why, in a given run? Eventually yes, I suppose.

File: testsuite/classlibrary/TestDocument.sc (hunk 2 above)
Comment (jamshark70) [reply testsuite/classlibrary/TestDocument.sc]: Normal way to handle asynchronicity is with a `condition.hang` in the async response (initAction here). But we are testing whether the initAction fired or not. Therefore there is some reasonable expectation that the initAction may fail to fire. In that case, the Condition never unhangs and voilà, the entire test suite has ground to a halt.

This situation is impossible to test without a timeout. So, please advise what is the interface provided by the UnitTest suite for asynchronicity with timeout, and I'll very happily update (as I've been waiting years for that interface).

File: testsuite/classlibrary/TestDocument.sc (hunk 3 above)
Comment (jamshark70) [reply testsuite/classlibrary/TestDocument.sc]: It was never clear to me what is the outcome of raising an exception in a test. What eventually happens upstream? I guess it makes sense if the test is reported as a failure due to exception.

File: testsuite/classlibrary/TestDocument.sc (hunk 1 above)
Comment (snappizz) [reply testsuite/classlibrary/TestDocument.sc]: just remembered there's an open ticket for this: #4384. i propose adding `// TODO: skip` to all skipped tests so we can easily grep for them later when test skipping is implemented.

File: testsuite/classlibrary/TestDocument.sc (hunk 3 above)
Comment (snappizz) [reply testsuite/classlibrary/TestDocument.sc]: i think the entire test suite exits. man that's annoying. i'd expect an exception to be pretty much the same as a failed assertion. no need for you to change anything here then.

File: testsuite/classlibrary/TestDocument.sc (hunk 2 above)
Comment (jamshark70) [reply testsuite/classlibrary/TestDocument.sc]: BTW I've started working on a Condition:hangWithTimeout so that this might not be a problem in the future.

Comment (snappizz) [reply testsuite/classlibrary/TestDocument.sc]: i don't work a lot with UnitTest, but i think it's `UnitTest:wait`? if that doesn't work, leaving in the `0.1.wait` is fine as long as there is a comment justifying it for anyone looking at the code.

Comment (jamshark70) [reply testsuite/classlibrary/TestDocument.sc]: Oh, but there is #3895. Well, what to do. I'll email the dev list.

Comment (snappizz) [reply testsuite/classlibrary/TestDocument.sc]: could you try `UnitTest:wait` and see if it works here?

**Comment by telephon** [comment]
>BTW I've started working on a Condition:hangWithTimeout so that this might not be a problem in the future.

There was a discussion about this in the past. I remember that the result was that you can just write `.hang(0.5)` or whatever time you want as timeout, and that it was a documentation issue.

https://github.com/supercollider/supercollider/pull/3300

**Comment by jamshark70** [comment]
> There was a discussion about this in the past. I remember that the result was that you can just write .hang(0.5) or whatever time you want as timeout, and that it was a documentation issue.

That discussion eventually rejected that solution.

```
(
r = fork {
	var c = Condition.new,
	resp = OSCFunc({ c.unhang }, '/test'),
	now;
	thisThread.clock.sched(0.25, {
		"let's pretend we got an async reply in 0.25 sec".postln;
		NetAddr.localAddr.sendMsg("/test");
	});
	c.hang(2);
	now = thisThread.beats;
	loop {
		(thisThread.beats - now).debug("beats since unhanging");
		1.0.wait;
	};
};
)

let's pretend we got an async reply in 0.25 sec
beats since unhanging: 0.0
beats since unhanging: 1.0
beats since unhanging: 1.74982049
beats since unhanging: 2.0
beats since unhanging: 2.74982049

r.stop;
```

File: testsuite/classlibrary/TestDocument.sc (hunk 1 above)
Comment (jamshark70) [reply testsuite/classlibrary/TestDocument.sc]: Done.

File: testsuite/classlibrary/TestDocument.sc (hunk 2 above)
Comment (jamshark70) [reply testsuite/classlibrary/TestDocument.sc]: Yeah, UnitTest:wait is better. I didn't know about that one.

Comment (jamshark70) [reply testsuite/classlibrary/TestDocument.sc]: BTW fixed 10 days ago. AFAICS this is ready to go?

File: testsuite/classlibrary/TestDocument.sc [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+				Document.initAction = save;
+			};
+			this.assert(success, "Document.new should fire Document.initAction function");
+		} {
+			// TODO: skip
+		};
+	}
+
+	test_open_autorun_document_runs_code {
+		var path = Platform.defaultTempDir +/+ "autoRunTest%.scd".format(UniqueID.next),
```
Comment (brianlheim) [inline testsuite/classlibrary/TestDocument.sc]: what's the intent here with `UniqueID.next`?

Comment (jamshark70) [reply testsuite/classlibrary/TestDocument.sc]: Hm, I think my intent was to reduce the likelihood of crashing into a leftover file in the temp directory. But UniqueID probably wouldn't work for that.

I'll just take it out.

**Pull request merged**